test/cpp/end2end/grpclb_end2end_test.cc

```diff
@@ -781,6 +785,52 @@ TEST_F(SingleBalancerTest, Vanilla) {
   EXPECT_EQ("grpclb", channel_->GetLoadBalancingPolicyName());
 }
 
+TEST_F(SingleBalancerTest, SubchannelCaching) {
+  ResetStub(/*fallback_timeout=*/0, /*expected_targets=*/"",
+            /*subchannel_cache_delay_ms=*/1500);
+  SetNextResolutionAllBalancers();
+  // Initially send all backends.
+  ScheduleResponseForBalancer(
+      0, BuildResponseForBackends(GetBackendPorts(), {}), 0);
+  // Then remove backends 0 and 1.
+  ScheduleResponseForBalancer(
+      0, BuildResponseForBackends(GetBackendPorts(2), {}), 1000);
+  // Now re-add backend 1.
+  ScheduleResponseForBalancer(
+      0, BuildResponseForBackends(GetBackendPorts(1), {}), 1000);
+  // Wait for all backends to come online.
+  WaitForAllBackends();
+  // Send RPCs for long enough to get all responses.
+  gpr_timespec deadline = grpc_timeout_milliseconds_to_deadline(3000);
+  do {
+    CheckRpcSendOk();
+  } while (gpr_time_cmp(gpr_now(GPR_CLOCK_MONOTONIC), deadline) < 0);
+  // Backend 0 should have received less traffic than the others.
+  // Backend 1 would have received less traffic than 2 and 3.
+  gpr_log(GPR_INFO, "BACKEND 0: %" PRIuPTR " requests",
+          backends_[0]->service_.request_count());
+  EXPECT_GT(backends_[0]->service_.request_count(), 0);
+  for (size_t i = 1; i < backends_.size(); ++i) {
+    gpr_log(GPR_INFO, "BACKEND %" PRIuPTR ": %" PRIuPTR " requests", i,
+            backends_[i]->service_.request_count());
+    EXPECT_GT(backends_[i]->service_.request_count(),
+              backends_[0]->service_.request_count())
+        << "backend " << i;
+    if (i >= 2) {
+      EXPECT_GT(backends_[i]->service_.request_count(),
+                backends_[1]->service_.request_count())
+          << "backend " << i;
+    }
+  }
+  // TODO(roth): This should ideally check that backend 1 never lost its
```

Comment:
I think we can check that backend 1 never lost its connection by checking that it only received RPCs from one peer IP:port

`EXPECT_EQ(1UL, backends_[1]->service_.clients().size());` - like above?

Reply:
Good idea.  Done.